korangar/src/loaders/animation/mod.rs

```diff
@@ -224,59 +222,55 @@ impl AnimationLoader {
+                let center_x = (frame.size.x - 1) / 2;
+                min_left = min(min_left, frame.offset.x - center_x);
+                max_right = max(max_right, frame.offset.x + (frame.size.x - 1 - center_x));
+
+                let center_y = (frame.size.y - 1) / 2;
+                min_top = min(min_top, frame.offset.y - center_y);
+                max_bottom = max(max_bottom, frame.offset.y + (frame.size.y - 1 - center_y));
+            });
 
+            // From here, the code start not using pixel coordinates (center of pixel
+            // coordinates), but instead pixel border + pixel center
+            // coordinates. This also means that we start using float
+            // numbers.
+            // |x|x|x| - x is the pixel center and | is the pixel border.
+            // `x` is in have number format of X.0 and `|`` is in the form X.5.
             animations[action_index].frames.iter_mut().for_each(|frame| {
                 frame.size = calculate_new_size(min_top, max_bottom, min_left, max_right);
-                // Set the origin point at (0, 0) correctly by applying an offset in y by
-                // max_bottom.
+                // Set the bottom part of the image as (0, 0), to shift easily after rotation
                 frame.offset = Vector2::new(0, -max_bottom);
 
                 #[cfg(feature = "debug")]
                 get_cross_matrix(frame);
                 for frame_part in frame.frame_parts.iter_mut() {
-                    let frame_top_left = frame.offset - (frame.size - Vector2::new(1, 1)) / 2;
-                    let frame_part_top_left = frame_part.offset - (frame_part.size - Vector2::new(1, 1)) / 2;
+                    let frame_size = frame.size.map(|val| val as f32);
+                    let frame_origin = frame.offset.map(|val| val as f32);
+
+                    // The offset is to move to pixel corner.
+                    let frame_top_left = Vector2::new(-frame.size.x as f32 / 2.0, -frame.size.y as f32 + 0.5);
+
+                    let frame_part_top_left_shift = -((frame_part.size - Vector2::new(1, 1)) / 2).map(|val| val as f32);
+                    let frame_part_offset = frame_part.offset.map(|val| val as f32);
+                    // The offset is to move from the pixel center to pixel corner.
+                    let frame_part_top_left =
+                        (frame_origin + frame_part_offset + frame_part_top_left_shift) - Vector2::<f32>::from_value(0.5);
 
                     // Generate the key points of the frame rectangle.
                     let texture_frame_center = Vector2::new(0.0, 1.0);
-
                     // Generate the key points of the frame part rectangle.
                     // In the variables, we removed the term frame_part,
                     // but we are still working with the frame part.
-                    let new_vector = frame_part.size;
+                    let new_vector = frame_part.size.map(|val| val as f32);

```

Comment:
```suggestion
                    let new_vector = frame_part.size.map(|value| value as f32);

```

Reply:
I created a function to change from Vector2<i32> to Vector2<f32>, I think it is better.